### PR TITLE
Mg/props ts

### PR DESCRIPTION
**Pull request by jeremy-rifkin**

### CODE REVIEW

File: lib/properties.ts [hunk 1]
```diff
@@ -28,24 +28,31 @@ import path from 'path';
 import _ from 'underscore';
 
 import {logger} from './logger';
+import {PropertyGetter, PropertyValue, Widen} from './properties.interfaces';
 import {toProperty} from './utils';
 
-let properties = {};
+let properties: Record<string, Record<string, PropertyValue>> = {};
```
Comment (mattgodbolt) [inline lib/properties.ts:34]: thanks for fixing up my overly excited types

File: static/panes/output.ts [hunk 2]
```diff
@@ -178,11 +178,7 @@ export class Output extends Pane<OutputState> {
             if (obj.text === '') {
                 this.add('<br/>');
             } else {
-                this.add(
-                    this.normalAnsiToHtml.toHtml(obj.text),
-                    lineNumber,
-                    columnNumber,obj.tag?.file
-                );
+                this.add(this.normalAnsiToHtml.toHtml(obj.text), lineNumber, columnNumber, obj.tag?.file);
```
Comment (mattgodbolt) [inline static/panes/output.ts:181]: unrelated?

File: static/styles/themes/dark-theme.scss [hunk 3]
```diff
@@ -128,7 +128,8 @@ textarea.form-control {
         background-color: #23272b !important;
     }
 
-    &:disabled, &:disabled:hover {
+    &:disabled,
```
Comment (mattgodbolt) [inline static/styles/themes/dark-theme.scss:131]: unrelated?

File: static/panes/output.ts (hunk 2 above)
Comment (jeremy-rifkin) [reply static/panes/output.ts:181]: Unrelated - I ran prettier

Comment (mattgodbolt) [reply static/panes/output.ts:181]: Got it! As you should! (though stupid precommits....) anyway, I'll check in the prettier first then you can merge...

File: lib/base-compiler.ts [hunk 4]
```diff
@@ -83,8 +84,8 @@ export class BaseCompiler {
     public lang: Language;
     protected compileFilename: string;
     protected env: any;
-    // Note that this can also return undefined and boolean | number, but we still need properties.js to ts
-    protected compilerProps: (key: string, defaultValue?: string) => string;
+    // Note that this can also return undefined and boolean | number, but we still need properties.ts to ts
```
Comment (mattgodbolt) [inline lib/base-compiler.ts:87]: nitpick (probably from my OG change...so I blame me) -> this comment makes no sense any more. Maybe just remove it?

File: lib/properties.interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export type PropertyValue = string | boolean | number | undefined;
```
Comment (mattgodbolt) [inline lib/properties.interfaces.ts:25]: Thanks for the significant witchcraft here....I will try and learn enough TS to understand it :)

File: lib/properties.ts [hunk 6]
```diff
@@ -28,24 +28,31 @@ import path from 'path';
...
 
-export function get(base, property, defaultValue) {
+export function get(base: string, property: string, defaultValue: undefined): PropertyValue;
+export function get<T extends PropertyValue>(
+    base: string,
+    property: string,
+    defaultValue: Widen<T>,
+): typeof defaultValue;
+export function get<T extends PropertyValue>(base: string, property: string, defaultValue?: unknown): T;
+export function get(base: string, property: string, defaultValue?: unknown): unknown {
```
Comment (mattgodbolt) [inline lib/properties.ts:55]: The final 👨🏻‍🍳 💋 would be to check the return value really is a `T` ... but... I don't know if that's possible in TS.

**Review by mattgodbolt — APPROVED** [review]
One comment and one wishlist (not to block anything though!) Thank you!

File: lib/properties.ts (hunk 6 above)
Comment (jeremy-rifkin) [reply lib/properties.ts:55]: I was thinking the same, it'd be really nice to enforce this. I think it is possible if we require a default argument always be passed, but I'm not sure how we'd handle cases were we expect, e.g., `string | undefined` from the Props.

Alternatively some day we could make the props interface look like this
```ts
props.getString("key");
props.getNumber("key");
props.getBool("key");
props.getOptionalString("key");
props.getOptionalNumber("key");
props.getOptionalBool("key");
```

File: lib/properties.interfaces.ts (hunk 5 above)
Comment (jeremy-rifkin) [reply lib/properties.interfaces.ts:25]: Huge thanks to the wizards on the typescript community discord server! I barely understand it myself!

File: lib/base-compiler.ts (hunk 4 above)
Comment (jeremy-rifkin) [reply lib/base-compiler.ts:87]: Good catch, I'll go ahead and remove the comment

**Comment by mattgodbolt** [comment]
woo!
